File: algorithms/graph/maximum-bipartite-matching/go/maximum_bipartite_matching.go
package maximumbipartitematching

func MaximumBipartiteMatching(arr []int) int {
	nLeft := arr[0]; nRight := arr[1]; m := arr[2]
	adj := make([][]int, nLeft)
	for i := 0; i < nLeft; i++ { adj[i] = []int{} }
	for i := 0; i < m; i++ { adj[arr[3+2*i]] = append(adj[arr[3+2*i]], arr[3+2*i+1]) }
	matchRight := make([]int, nRight)
	for i := range matchRight { matchRight[i] = -1 }

	var dfs func(u int, visited []bool) bool
	dfs = func(u int, visited []bool) bool {
		for _, v := range adj[u] {
			if !visited[v] {
				visited[v] = true
				if matchRight[v] == -1 || dfs(matchRight[v], visited) {
					matchRight[v] = u; return true
				}
			}
		}
		return false
	}

	result := 0
	for u := 0; u < nLeft; u++ {
		visited := make([]bool, nRight)
		if dfs(u, visited) { result++ }
	}
	return result
}

File: algorithms/graph/maximum-bipartite-matching/python/maximum_bipartite_matching.py
def maximum_bipartite_matching(arr: list[int]) -> int:
    n_left = arr[0]
    n_right = arr[1]
    m = arr[2]
    adj = [[] for _ in range(n_left)]
    for i in range(m):
        u = arr[3 + 2 * i]
        v = arr[3 + 2 * i + 1]
        adj[u].append(v)

    match_right = [-1] * n_right

    def dfs(u, visited):
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                if match_right[v] == -1 or dfs(match_right[v], visited):
                    match_right[v] = u
                    return True
        return False

    result = 0
    for u in range(n_left):
        visited = [False] * n_right
        if dfs(u, visited):
            result += 1
    return result
